flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -0,0 +1,1000 @@
+					if (exception != null && asyncException == null) {
+						asyncException = exception;
+					}
+					acknowledgeMessage();
+				}
+			};
+		}
+
+		super.open(configuration);
+	}
+
+	@Override
+	public void invoke(KafkaTransactionState transaction, IN next) throws Exception {
+		invokeInternal(transaction, next, Long.MAX_VALUE);
+	}
+
+	private void invokeInternal(KafkaTransactionState transaction, IN next, long elementTimestamp) throws Exception {
+		checkErroneous();
+
+		byte[] serializedKey = schema.serializeKey(next);
+		byte[] serializedValue = schema.serializeValue(next);
+		String targetTopic = schema.getTargetTopic(next);
+		if (targetTopic == null) {
+			targetTopic = defaultTopicId;
+		}
+
+		Long timestamp = null;
+		if (this.writeTimestampToKafka) {
+			timestamp = elementTimestamp;
+		}
+
+		ProducerRecord<byte[], byte[]> record;
+		int[] partitions = topicPartitionsMap.get(targetTopic);
+		if (null == partitions) {
+			partitions = getPartitionsByTopic(targetTopic, transaction.producer);
+			topicPartitionsMap.put(targetTopic, partitions);
+		}
+		if (flinkKafkaPartitioner == null) {
+			record = new ProducerRecord<>(targetTopic, null, timestamp, serializedKey, serializedValue);
+		} else {
+			record = new ProducerRecord<>(targetTopic, flinkKafkaPartitioner.partition(next, serializedKey, serializedValue, targetTopic, partitions), timestamp, serializedKey, serializedValue);
+		}
+		pendingRecords.incrementAndGet();
+		transaction.producer.send(record, callback);
+	}
+
+	@Override
+	public void close() throws Exception {
+		if (currentTransaction != null) {
+			// to avoid exceptions on aborting transactions with some pending records
+			flush(currentTransaction);
+		}
+		try {
+			super.close();
+		}
+		catch (Exception e) {
+			asyncException = ExceptionUtils.firstOrSuppressed(e, asyncException);
+		}
+		try {
+			producersPool.close();
+		}
+		catch (Exception e) {
+			asyncException = ExceptionUtils.firstOrSuppressed(e, asyncException);
+		}
+		// make sure we propagate pending errors
+		checkErroneous();
+	}
+
+	// ------------------- Logic for handling checkpoint flushing -------------------------- //
+
+	@Override
+	protected KafkaTransactionState beginTransaction() throws Exception {
+		switch (semantic) {
+			case EXACTLY_ONCE:
+				FlinkKafkaProducer<byte[], byte[]> producer = producersPool.poll();
+				if (producer == null) {
+					String transactionalId = availableTransactionalIds.poll();
+					if (transactionalId == null) {
+						throw new Exception(
+							"Too many ongoing snapshots. Increase kafka producers pool size or decrease number of concurrent checktpoins.");
+					}
+					usedTransactionalIds.add(transactionalId);
+					producer = initTransactionalProducer(transactionalId, true);
+					producer.initTransactions();
+				}
+				producer.beginTransaction();
+				return new KafkaTransactionState(producer.getTransactionalId(), producer);
+			case AT_LEAST_ONCE:
+			case NONE:
```

Comment:
Because `NONE` and `AT_LEAST_ONCE` do not need to start new transaction, the `beginTransaction()` implementation for both them is blank. Those two differ only in `preCommit()` where the later one flushes the data.